system/jlib/jscm.hpp

```diff
@@ -94,6 +109,80 @@ template <class CLASS> class Shared
     CLASS * ptr;
 };
 
+// Similar to class Shared<X>, but thread safe versions of the functions (avoid the need for critical sectiions)
+template <class CLASS> class AtomicShared
+{
+public:
+    inline AtomicShared()                              { ptr.store(nullptr, std::memory_order_relaxed); }
+    inline AtomicShared(CLASS * _ptr, bool owned)      { ptr.store(_ptr, std::memory_order_relaxed); if (!owned && _ptr) _ptr->Link(); }
+    inline AtomicShared(const AtomicShared & other)    { ptr.store(other.getLink(), std::memory_order_relaxed); }
+#if defined(__cplusplus) && __cplusplus >= 201100
+    inline AtomicShared(AtomicShared && other)         { ptr.store(other.getLink(), std::memory_order_relaxed); }
+#endif
+    inline ~AtomicShared()                             { ::Release(ptr.load(std::memory_order_relaxed)); }
+    inline AtomicShared<CLASS> & operator = (const AtomicShared<CLASS> & other) { this->setown(other.getLink()); return *this;  }
+
+    inline void clear()                                { ::Release(getClear()); }
+    inline CLASS * getClear()
+    {
+        return ptr.exchange(nullptr);
+    }
+    //Slightly dubious whether this should be included - not valid if clear() could be called concurrently
+    inline CLASS * getLink() const
+    {
+        CLASS * temp = ptr;
+        if (temp)
+            temp->Link();
+        return temp;
+    }
+    inline void set(CLASS * _ptr)
+    {
+        if (ptr != _ptr)
+        {
+            LINK(_ptr);
+            this->setown(_ptr);
+        }
+    }
+    inline bool setownNull(CLASS * _ptr)
+    {
+        CLASS * expected = nullptr;
+        if (ptr.compare_exchange_strong(expected, _ptr))
+            return true;
+        ::Release(_ptr);
+        return false;
+    }
+    inline void set(const AtomicShared<CLASS> &other)
+    {
+        this->setown(other.getLink());
+    }
+    inline void setown(CLASS * _ptr)
+    {
+        CLASS * temp = ptr.exchange(_ptr);
+        ::Release(temp);
+    }
+    inline CLASS * swap(CLASS * _ptr)
+    {
+        return ptr.exchange(_ptr);
+    }
+    inline void swap(AtomicShared<CLASS> & other)
+    {
+        CLASS * temp = other.getClear();
```

Comment:
Should be this->getLink(), or do the other way around.

 and should define the semantics so that this is only ever updated once.

Reply:
Now changed and added a comment to document